change bit_blt call into print_window

bensandland opened the pull request:
Currently it is not possible to take screenshots of certain windows (non Windows native or hardware accelerated), eg. Google Chrome - when using bit_blt the image created is black.

I changed the usage of this function into print_window and it is now possible to take screenshots of whole webpages in Chrome, using the pid/hwnd for the process/window

I have ran all specs and all of them passed on my end. Let me know if you want anything else changed?

bensandland commented:
I have tested further and I can see it no longer works when trying to capture the desktop. I am working on a fix for this, and also trying to make it possible to capture the desktop across multiple monitors. It might not be possible with the print_window function, in that case I will try to make an abstraction which uses either bit_blt or print_window depending on the context.

jarmo commented:
Hi, that's cool!

Can you explain what kind of backwards incompatibility issues these changes might bring? Does it also work on older Windows'es like XP for example or is `print_window` an older thing?

jarmo commented on lib/win32/screenshot/bitmap_maker.rb:
```diff
@@ -74,14 +102,34 @@ def capture_area(hwnd, context, x1, y1, x2, y2)
           release_dc(0, hScreenDC)
         end
 
+        def desktop_dimensions
```
Maybe extract desktop related methods into a separate Struct type class:
```
# desktop.rb

Win32::Screenshot::Desktop = Struct.new(:left, :top, :width, :height) do
  def dimensions
    [left, top, width, height]
  end
end
```

bensandland replied:
Done. Although the test for capturing desktop lost a little clarity: [link](https://github.com/bensandland/win32screenshot/blob/master/spec/win32/screenshot/take_spec.rb#L34)

Is it worthwhile to add getter functions to for width and height specifically?

jarmo replied:
Have this method return `Desktop` Struct instead of its dimensions. Then whoever needs dimensions can do `desktop.dimensions` and whoever needs height/width or whatever from the Struct can do just like that (e.g `desktop.height` or `desktop.width`. Struct already has getters for all of its fields.

jarmo replied:
Of course rename this method to `desktop` too instead of `desktop_dimensions`.

bensandland replied:
Ah yes of course, that makes sense. Changed it now

bensandland commented:
Hey.

I will admit this is not something I have considered too much, but looking at the [Microsoft docs](https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-printwindow) it looks like the "Minimum Supported Client" is Windows XP, although it specifies for desktop apps only.

However it looks like the PW_RENDERFULLCONTENT constant was introduced in Windows 8.1 according to this [SO comment](https://stackoverflow.com/a/40042587)

Is there a certain way you usually test for backward combability? In a VM perhaps?

Nakilon commented:
Have you benchmarked this?

jarmo commented on lib/win32/screenshot/bitmap_maker.rb:
```diff
@@ -40,31 +42,57 @@ class << self
         def capture_all(hwnd, context)
           width, height = dimensions_for(hwnd, context)
-          capture_area(hwnd, context, 0, 0, width, height)
+          capture_area(hwnd, context, width: width, height: height)
```
What's the point of this change of having `width: width` and `height: height`?

I have to admit that I have not written newer Ruby syntax and I'm not aware of all of its features, I would have expected that it is just passing a Hash, but looking at the method signature of `capture_area` it seems like keyword arguments? Maybe not add that extra complexity?

jarmo commented on lib/win32/screenshot/bitmap_maker.rb:
```diff
@@ -40,31 +42,57 @@ class << self
 
-        SRCCOPY = 0x00CC0020
-        DIB_RGB_COLORS = 0
+        def capture_area(hwnd, context, width:, height:)
```
Look previous comment.

bensandland commented on lib/win32/screenshot/take.rb:
```diff
@@ -80,13 +84,21 @@ def window(opts)
+        def screenshot_area(hwnd, opts, window = true)
+          validate_coordinates(hwnd, opts[:context], *opts[:area])
+          if window
+            BitmapMaker.capture_area(hwnd, opts[:context], width: opts[:area][2], height: opts[:area][3])
```
What do you think about something like [this?](https://github.com/bensandland/win32screenshot/commit/949f911c8ca22ed224de9b47e3be4fe3dd6b15be#diff-0b14d6ff697b973f88b7bc8dc991aaaee51b711f91f368b52157fccbc35b5cc7L96)

jarmo commented on spec/win32/screenshot/take_spec.rb:
```diff
@@ -31,15 +31,15 @@
-    hwnd = Win32::Screenshot::BitmapMaker.desktop_window
-    expect([image.width, image.height]).to eq(Win32::Screenshot::BitmapMaker.dimensions_for(hwnd, :window))
+
+    expect([image.width, image.height]).to eq([Win32::Screenshot::BitmapMaker::desktop.width, Win32::Screenshot::BitmapMaker::desktop.height])
```
Take desktop into local variable - easier to read then.

jarmo replied:
@bensandland this is still undone.

jarmo commented on spec/win32/screenshot/take_spec.rb:
```diff
@@ -93,8 +93,8 @@
     save_and_verify_image(image, 'notepad_area')
-    expect(image.width).to eq(70)
-    expect(image.height).to eq(120)
+    expect(image.width).to eq(100)
```
And again about dimensions.

jarmo commented on lib/win32/screenshot/take.rb:
```diff
@@ -80,13 +84,22 @@ def window(opts)
+        def screenshot_area(hwnd, opts, window = true)
+          validate_coordinates(hwnd, opts[:context], *opts[:area])
+          if window
+            width, height = opts[:area].last(2)
```
I'm afraid that this is the line culprit of changing dimensions in tests.

bensandland commented on spec/win32/screenshot/take_spec.rb:
```diff
@@ -19,8 +19,8 @@
     save_and_verify_image(image, 'foreground_area')
-    expect(image.width).to eq(70)
-    expect(image.height).to eq(120)
+    expect(image.width).to eq(100)
```
I believe the culprit are that these lines from the original [capture_area](https://github.com/jarmo/win32screenshot/blob/88e209986d1de09f5afff0d1f4a9bcfbdf281db6/lib/win32/screenshot/bitmap_maker.rb#L54) function is no longer there.

I'll try to figure out a better solution for my current implementation

jarmo replied:
Maybe since your additions capture the whole window (even not just the visible area, as was done before), then it makes sense to add a separate :type to the `Take#of` method to preserve backwards compatibility? Maybe something in the lines of `:window_full` or something to hint that it is different from `:window`?

I'm pretty sure that there exists code where `:area` has been set and when `x1` and/or `y1` are different from 0 then behavior will change with these changes (and even having these options do not add any value when they are not used).

It would be great though to detect somehow if capturing a screenshot with `:window` is even possible or if black screen will be captured to throw an error and suggest using `:window_full` or whatever. Not sure if it's possible.

Now that I'm thinking about it when writing this comment I'm starting to think that maybe backwards-incompatibility should be introduced and `:area` option should be removed altogether - let's just try to make full screenshot all the time from whatever is asked (either `:desktop`, `:foreground` or `:window`) and if anyone cares only about the area then image postprocessing should be done separately by some other library/tool. In that way `win32screenshot` does only one thing - takes screenshots.

What are your thoughts?

bensandland replied:
I think that makes a lot of sense. As it is right now, our team is only interested in capturing windows/whole desktop. Maybe I will work on something like specifying primary monitor when capturing desktop.

I'll work on committing some code that removes the area functionality all together. Thanks for your cooperation

jarmo replied:
Please also remove mentions of `area` options in the documentation of `Take#of`.

bensandland commented:
Why is [this](https://github.com/bensandland/win32screenshot/blob/4e435542e101e9b1f745eb39309ff9b926708871/lib/win32/screenshot/take.rb#L67) line needed? As soon as I remove it the win.active? call causes RAutomation to raise a NoMethodError. 

I have also been wondering about the inclusion of this logic regarding [window dimensions.](https://github.com/bensandland/win32screenshot/blob/4e435542e101e9b1f745eb39309ff9b926708871/lib/win32/screenshot/bitmap_maker.rb#L121) Is there a reason for this addtion to the dimensions? It is causing the screenshots to contain a small border - we are using screenshots for image recognition and this added border could cause some issues regarding offsets.

jarmo commented:
> Why is [this](https://github.com/bensandland/win32screenshot/blob/4e435542e101e9b1f745eb39309ff9b926708871/lib/win32/screenshot/take.rb#L67) line needed? As soon as I remove it the win.active? call causes RAutomation to raise a NoMethodError.

Not sure. Maybe RAutomation also has something to do with `context` and gets confused when this is passed. But why do you need/want to remove that line?

> I have also been wondering about the inclusion of this logic regarding [window dimensions.](https://github.com/bensandland/win32screenshot/blob/4e435542e101e9b1f745eb39309ff9b926708871/lib/win32/screenshot/bitmap_maker.rb#L121) Is there a reason for this addition to the dimensions? It is causing the screenshots to contain a small border - we are using screenshots for image recognition and this added border could cause some issues regarding offsets.

Do you mean this `+ 1` logic? Not sure to be honest. Maybe it was meant to cut off window border at the time by assuming that border width/height is 1px?

bensandland commented:
> > Why is [this](https://github.com/bensandland/win32screenshot/blob/4e435542e101e9b1f745eb39309ff9b926708871/lib/win32/screenshot/take.rb#L67) line needed? As soon as I remove it the win.active? call causes RAutomation to raise a NoMethodError.
> 
> Not sure. Maybe RAutomation also has something to do with `context` and gets confused when this is passed. But why do you need/want to remove that line?
> 

I was trying to remove that line since I noticed screenshots were the same when using :window / :client as context, except for the "+1 logic" of course. Not sure I am using it as intended?

Eg. I was trying to take a screenshot of a File explorer window, like so:
`Win32::Screenshot::Take.of(:window, title: /new folder/i).write!("file_explorer.png")`
`Win32::Screenshot::Take.of(:window, context: :client, title: /new folder/i).write!("file_explorer_client.png")`

I have attached the screenshots - there is a small difference when using :client but it just looks like the screenshot of the a window is a little inaccurate?

![file_explorer_client](https://user-images.githubusercontent.com/33618088/164005381-dd483822-c688-4a58-a4f9-6fe71cc1f830.png)
![file_explorer](https://user-images.githubusercontent.com/33618088/164005385-4385555d-3be5-4196-9858-3a94a965a259.png)

> > I have also been wondering about the inclusion of this logic regarding [window dimensions.](https://github.com/bensandland/win32screenshot/blob/4e435542e101e9b1f745eb39309ff9b926708871/lib/win32/screenshot/bitmap_maker.rb#L121) Is there a reason for this addition to the dimensions? It is causing the screenshots to contain a small border - we are using screenshots for image recognition and this added border could cause some issues regarding offsets.
> 


> Do you mean this `+ 1` logic? Not sure to be honest. Maybe it was meant to cut off window border at the time by assuming that border width/height is 1px?

How would you feel about removing this piece of code?

bensandland commented:
I tried to change the logic in [dimensions_for](https://github.com/bensandland/win32screenshot/blob/4e435542e101e9b1f745eb39309ff9b926708871/lib/win32/screenshot/bitmap_maker.rb#L121)

To this:

         def dimensions_for(hwnd, context)
            rect = [0, 0, 0, 0].pack('l4')
            BitmapMaker.send("#{context}_rect", hwnd.to_i, rect)
            left, top, width, height = rect.unpack('l4')

            [width - left, height - top]
         end

These are the screenshots it created:

:window
![new_file_explorer](https://user-images.githubusercontent.com/33618088/164008726-d63d58db-56b6-4bc6-a24e-c3f9a16167b0.png)
:client
![new_file_explorer_client](https://user-images.githubusercontent.com/33618088/164008729-d93e2119-3a2d-4be1-8ed2-5f2dd0a2fb0a.png)

So it looks like the same issue is present here, where the screenshot is skewed slightly towards the left - as you can see in the image above there is a bit of the window missing in the right of the screenshot

jarmo commented:
> So it looks like the same issue is present here, where the screenshot is skewed slightly towards the left - as you can see in the image above there is a bit of the window missing in the right of the screenshot

Can we remove the `:client` option altogether and just keep `:window`? If `:window` is the only option left (don't have time to check right now) then move passing that option altogether. What do you think? Do you see any issues with that change?

bensandland commented:
> Can we remove the `:client` option altogether and just keep `:window`? If `:window` is the only option left (don't have time to check right now) then move passing that option altogether. What do you think? Do you see any issues with that change?

I don't see a problem with removing it. I have pushed the code to reflect this change.

I also cleaned up the tests in take_spec.rb, and added a new which utilizes the :class option. I had to hook add a new function in bitmap_maker to achieve this though.

All tests are passing on my end, if you have any issues with the new changes let me know.

bensandland commented:
@Nakilon I have added some benchmark results to branch [https://github.com/bensandland/win32screenshot/tree/benchmark](https://github.com/bensandland/win32screenshot/tree/benchmark) if you are interested

jarmo commented:
1) Am I right that `client_dc` and `client_rect` in `BitmapMaker` are not used anymore? Remove those lines then.

2) Do I understand correctly that this `class_name_for` test was added because you noticed that this test does not exist yet? Instead of adding a method into internal `BitmapMaker` API, which is only used for tests then I would try to get that `class_name` via using `RAutomation` like in the README. Something like this might work (untested, might not work as expected):
```
RAutomation::Window.new(hwnd: window.hwnd).class_names.first
```

And then delete that `class_name_for` method which is only used by tests.

bensandland commented:
> 1. Am I right that `client_dc` and `client_rect` in `BitmapMaker` are not used anymore? Remove those lines then.

You're right, I'll remove them.

> 2. Do I understand correctly that this `class_name_for` test was added because you noticed that this test does not exist yet? Instead of adding a method into internal `BitmapMaker` API, which is only used for tests then I would try to get that `class_name` via using `RAutomation` like in the README. Something like this might work (untested, might not work as expected):
> 
> ```
> RAutomation::Window.new(hwnd: window.hwnd).class_names.first
> ```
> 
> And then delete that `class_name_for` method which is only used by tests.

I tried doing it like this, but it looks like RAutomation returns the class name for the controls, instead of the class name for the window.

Eg. notepad 
`RAutomation::Window.new(:pid => @notepad).class_names.first`
This returns "Edit"

`window.class_names `
returns:  ["Edit", "msctls_statusbar32"]

This line of code from the test returns "Notepad"
`Win32::Screenshot::BitmapMaker.class_name_for(window.hwnd)`

I don't think anyone in the team I'm in wants to use this functionality, but I saw the test in the take_spec file and looked into how to fix it. Would you rather add this functionality to the RAutomation gem, or do you think we could drop it altogether? Maybe there already is a way of doing this without adding a new ffi hook, I'll investigate it a bit further.

jarmo commented:
@bensandland I think if no-one wants to use it then just remove it for simplicity.

bensandland commented:
> Almost everything looks good, but I got this late thought - why do we have `top` and `left` for `Desktop` struct? What are they representing there? Aren't they always with a value of 0? If that's the case then I would remove `top` and `left` since they might send out a mixed signal of being something different from 0, but I can't think of any logical reason for that.
> 
> Can you elaborate?

That depends on the users desktop configuration. In my case I have 2 monitors, which means my top and left values are 0 and -1920. 

I don't know if my left value always will be 0, I'll have to do some more testing on other monitor configurations to confirm this. 

For more information regarding these values check out https://docs.microsoft.com/en-us/windows/win32/api/winuser/nf-winuser-getsystemmetrics - the ones in question are SM_XVIRTUALSCREEN and SM_YVIRTUALSCREEN.

jarmo commented:
@bensandland after some further googling about `SM_XVIRTUALSCREEN` and `SM_YVIRTUALSCREEN` it seems to me that they are relative coordinates from primary monitor. This means, according to my understanding, if for example your secondary monitor is position right-side of the primary monitor then both will return 0 as their value. However, if you move your secondary monitor to top left of your primary monitor then there will be negative numbers coming out from that function.

Can you confirm this? If that's the case then do a sanity check if anything goes wrong in `win32screenshot` when `top` and `left` will have negative values.

bensandland commented:
> @bensandland after some further googling about `SM_XVIRTUALSCREEN` and `SM_YVIRTUALSCREEN` it seems to me that they are relative coordinates from primary monitor. This means, according to my understanding, if for example your secondary monitor is position right-side of the primary monitor then both will return 0 as their value. However, if you move your secondary monitor to top left of your primary monitor then there will be negative numbers coming out from that function.
> 
> Can you confirm this? If that's the case then do a sanity check if anything goes wrong in `win32screenshot` when `top` and `left` will have negative values.

Alright yes, that makes a lot of sense. 
I have tried to do some experimenting like you said, and there seems to be no issues with left and top values being negative.
I setup my monitors like you described, having my secondary monitor placed at the top left of my main monitor.

Take a look at the screenshots I have attached and let me know if this is what you meant.

Monitor setup with secondary positioned top left of primary:

![monitor_setup](https://user-images.githubusercontent.com/33618088/173774283-c46b1c6e-1a96-4894-a05b-c389d7200b63.PNG)

Image of my full desktop while debugging, to showcase that function returns top and left values being negative:

![desktop](https://user-images.githubusercontent.com/33618088/173774279-197bdb03-ba0f-43e1-bb68-e906e3c5b4a2.png)

jarmo commented:
@bensandland yes, that is what I meant.

Merged changes and released `win32screenshot` 4.0.0. Thank you for your contribution and patience.
